### PR TITLE
Ex5

### PR DESCRIPTION
**Pull request by zyahav**

### CODE REVIEW

File: ex5B/ex5.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+function firstUniqueChar(str) {
+
+	var doNotLookForCharList = '', searchChar = ''; 
+
+	for (var searchIndex = 0; searchIndex < str.length; searchIndex++) {
+
+		searchChar = str[searchIndex];
+
+		//  Let's see if we did look for searchChar
+		if (doNotLookForCharList.includes(searchChar) != true) {
```
Comment (myok12) [inline ex5B/ex5.js]: Going back to my question, I think you are searching for the char both ahead (using the indexOf below) and backwards (in doNotLookForCharList), which is redundant. Put an `else` to the first `if` and see if it ever gets called.

File: ex5B/ex5Spec.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+describe("firstUniqueChar", function() {
+	it("finds another unique character in the string", function() {
+		expect(ex5.firstUniqueChar("abcdedcba")).toBe("e");
+	});
+
+	it("return null, we do not have a uniq char on string", function() {
+		expect(ex5.firstUniqueChar("abcdeedcba")).toBeNull();
+	});
+
+	it("finds another unique character in the strin", function() {
```
Comment (myok12) [inline ex5B/ex5Spec.js]: Ahinu, please install a vim plugin for dictionary. `strin` --> `string`

File: ex5B/ex5Spec.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+		expect(ex5.firstUniqueChar("abababcab")).toBe("c");
+	});
+
+	it("return null, we do not have a uniq char on strin", function() {
```
Comment (myok12) [inline ex5B/ex5Spec.js]: You have to have the test methods using different names, otherwise, you wouldn't know which one is broken when one does break. Change this one to:
`return null when we only have one [non-unique] char`

File: ex5B/ex5Spec.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+		expect(ex5.firstUniqueChar("aaaaaaaaa")).toBeNull();
+	});
+
+	it("works for a single characte", function() {
```
Comment (myok12) [inline ex5B/ex5Spec.js]: s/characte/character/

Comment (zyahav) [inline ex5B/ex5Spec.js]: done

File: ex5B/ex5.js (hunk 1 above)
Comment (zyahav) [inline ex5B/ex5.js]: firstUniqueChar("abcdeedcbaf");


VM924:13 doNotLookForCharList.includes(searchChar) false a
VM924:13 doNotLookForCharList.includes(searchChar) false b
VM924:13 doNotLookForCharList.includes(searchChar) false c
VM924:13 doNotLookForCharList.includes(searchChar) false d
VM924:13 doNotLookForCharList.includes(searchChar) false e
VM924:13 doNotLookForCharList.includes(searchChar) true e
VM924:13 doNotLookForCharList.includes(searchChar) true d
VM924:13 doNotLookForCharList.includes(searchChar) true c
VM924:13 doNotLookForCharList.includes(searchChar) true b
VM924:13 doNotLookForCharList.includes(searchChar) true a
VM924:13 doNotLookForCharList.includes(searchChar) false f
"f"

File: ex5B/ex5Spec.js (hunk 3 above)
Comment (zyahav) [inline ex5B/ex5Spec.js]: i have no idea 

	it("finds another unique character in the string", function() {
		expect(ex5.firstUniqueChar("abcdedcba")).toBe("e");
	});

	it("finds another unique character in the string", function() {
		expect(ex5.firstUniqueChar("abababcab")).toBe("c");
	});

Comment (myok12) [inline ex5B/ex5Spec.js]: If you don't have anything smart to say about the different cases, simply rename one slightly. Back in the day we used to put some random number, but probably just a slight English change is good, so change one to:
`finds some unique character in the string`

File: ex5B/ex5.js (hunk 1 above)
Comment (myok12) [inline ex5B/ex5.js]: Try to comment out the `if (doNotLookForCharList.includes(searchChar) != true) {` and see if your code still works.

Comment (zyahav) [inline ex5B/ex5.js]: Will do it together

File: ex5B/ex5Spec.js [hunk 5]
```diff
@@ -1,23 +1,23 @@
-var ex5 = require('./ex5.js');
+var firstUniqueChar = require('./ex5.js');
 
-describe("firstUniqueChar", function() {
-	it("finds another unique character in the string", function() {
-		expect(ex5.firstUniqueChar("abcdedcba")).toBe("e");
+describe("function firstUniqueChar", function() {
+	it("finds another unique character e in the string", function() {
```
Comment (myok12) [inline ex5B/ex5Spec.js]: The e is irrelevant for the test name -- it doesn't matter what letter is set to find. The title of the test should not have `e` in it.

File: ex5B/ex5.js (hunk 1 above)
Comment (zyahav) [inline ex5B/ex5.js]: Done

File: ex5B/ex5.js [hunk 6]
```diff
@@ -10,9 +10,46 @@
...
-
 function firstUniqueChar(str) {
 
+        var charList = {}, firstUniqueCharVal = null;
+
+        for (var searchIndex = 0; searchIndex < str.length; searchIndex++) {
+
+                if (charList[str.charAt(searchIndex)] === undefined) {
+
+			// Creat new property
```
Comment (myok12) [inline ex5B/ex5.js]: Didn't you install a dictionary? Creat-->Create

File: ex5B/ex5.js [hunk 7]
```diff
@@ -10,9 +10,46 @@
  * firstUniqueChar("aaaaaaaaa") === null
  * firstUniqueChar("a") === "a"
  */
-
 function firstUniqueChar(str) {
 
+        var charList = {}, firstUniqueCharVal = null;
```
Comment (myok12) [inline ex5B/ex5.js]: `charList` is no longer a list. rename to `charCounter`.

File: ex5B/ex5.js [hunk 8]
```diff
@@ -10,9 +10,46 @@
...
+			charList[str.charAt(searchIndex)] = 1;
+
+                } else {
+			// Add to property
+                        charList[str.charAt(searchIndex)] += 1;
+
+                }
+        }
+
+        for (searchIndex = 0; searchIndex < Object.keys(charList).length; searchIndex++) {
```
Comment (myok12) [inline ex5B/ex5.js]: You don't know how much work `Object.keys(charList)` require. Please cache this value into a variable (it will also read better).

File: ex5B/ex5.js [hunk 9]
```diff
@@ -10,9 +10,46 @@
...
+
+                if (charList[Object.keys(charList)[searchIndex]] === 1) {
+
+                        if (firstUniqueCharVal === null) {
+
+                                firstUniqueCharVal = Object.keys(charList)[searchIndex];
+
+                       } else {
+
+                                firstUniqueCharVal = null;
```
Comment (myok12) [inline ex5B/ex5.js]: I think it's probably will be easier to understand if you simply `return null` here instead of resetting the `firstUniqueCharVal` to null, then break, and return it. Then, you can also rename `firstUniqueCharVal` to `firstChar` (which is simpler to understand what it does)

File: ex5B/ex5Spec.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+	});
+
+	it("finds another unique character e in the string", function() {
+		expect(firstUniqueChar("abababcab")).toBe("c");
+	});
+
+	it("return null, all chars are same", function() {
+		expect(firstUniqueChar("aaaaaaaaa")).toBeNull();
+	});
+	it("works on really long strings", function() {
```
Comment (myok12) [inline ex5B/ex5Spec.js]: How long did it take before? How long does it take now?

File: ex5B/ex5Spec.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
+var firstUniqueChar = require('./ex5.js');
+
+describe("function firstUniqueChar", function() {
+	it("finds another unique character e in the string", function() {
+		expect(firstUniqueChar("abcdedcba")).toBe("e");
+	});
+
+	it("return null when we only have one [non-unique] char", function() {
```
Comment (myok12) [inline ex5B/ex5Spec.js]: This title is from an old test. It should say `return null, all chars are duplicated`

File: ex5B/ex5Spec.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+		for (i=100;i<100000;i++) {
+			longString += String.fromCharCode(i);
+		}
+		firstUniqueChar(longString);
+	});
+
+	it("works for a single character", function() {
+		expect(firstUniqueChar("a")).toBe("a");
+	});
+});
```
Comment (myok12) [inline ex5B/ex5Spec.js]: Please add a few tests when you have more than one non-repeating characters, and you will see they don't work as expected (they don't always return the *first* one).

File: ex5B/ex5.js (hunk 7 above)
Comment (zyahav) [inline ex5B/ex5.js]: Done

File: ex5B/ex5.js (hunk 9 above)
Comment (zyahav) [inline ex5B/ex5.js]: Done

Comment (zyahav) [inline ex5B/ex5.js]: why do need to use break???

File: ex5B/ex5Spec.js (hunk 11 above)
Comment (zyahav) [inline ex5B/ex5Spec.js]: Done

File: ex5B/ex5Spec.js (hunk 10 above)
Comment (zyahav) [inline ex5B/ex5Spec.js]: Old time 6s to 16s
New time 0.34s

File: ex5B/ex5Spec.js (hunk 12 above)
Comment (zyahav) [inline ex5B/ex5Spec.js]: Done

Comment (zyahav) [inline ex5B/ex5Spec.js]: ????

File: ex5B/ex5.js [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+ * should return null.
+ *
+ * For example:
+ * firstUniqueChar("abcdedcba") === "e"
+ * firstUniqueChar("abcdeedcba") === null
+ * firstUniqueChar("abababcab") === "c"
+ * firstUniqueChar("aaaaaaaaa") === null
+ * firstUniqueChar("a") === "a"
+ */
+function firstUniqueChar(str) {
```
Comment (myok12) [inline ex5B/ex5.js]: You don't necessarily need to change this, but when you have a blank line between every two lines, it loses its meaning (you can just increase you editor's line height to get a similar visual effect). Use it more sparingly, where it adds some insight.

File: ex5B/ex5.js [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+  var objCharCounter = {}, objFirstChar = { charVal: null }, searchIndex = 0;
+
+  for (searchIndex = 0; searchIndex < str.length; searchIndex++) {
+
+    if (objCharCounter[str.charAt(searchIndex)] === undefined) {
```
Comment (myok12) [inline ex5B/ex5.js]: Running `str.charAt(searchIndex)` takes time. And you are running it twice in any case. So, instead, cache it into a variable: `var currentChar = str.charAt(searchIndex)`

File: ex5B/ex5.js [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+      // Create new property
+      objCharCounter[str.charAt(searchIndex)] = { index: searchIndex, count: 1};
```
Comment (myok12) [inline ex5B/ex5.js]: While the property `index` is fine, it is not really accurate when you have `count >1`. I would rename it to `firstIndex`

File: ex5B/ex5.js [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+    } else {
+
+      // Add to property
+      objCharCounter[str.charAt(searchIndex)].count += 1;
+
+   }
+
+  }
+
+  charArray = Object.keys(objCharCounter);
```
Comment (myok12) [inline ex5B/ex5.js]: You can see that this part is totally unrelated to the one above. I would suggest breaking this (and actually the one above too) into separate smaller functions (`buildCharIndicesMap` and `findFirstUniqueCharInMap`).

Comment (myok12) [inline ex5B/ex5.js]: I would also move the `var` (or better `let`) (unless it conflicts with the style guide) because it confuses when you read it at the top, because it's not being used there.

File: ex5B/ex5.js [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+  for (searchIndex = 0; searchIndex < Object.keys(objCharCounter).length; searchIndex++) {
```
Comment (myok12) [inline ex5B/ex5.js]: `searchIndex < Object.keys(objCharCounter).length` gets checked with every iteration of the loop and because `Object.keys(objCharCounter)` might take many CPU cycles, you should cache it into a variable (this could be why your code runs slower than before).

File: ex5B/ex5.js (hunk 16 above)
Comment (myok12) [inline ex5B/ex5.js]: Move the objFirstChar too.

File: ex5B/ex5.js [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+
+		if (objCharCounter[charArray[searchIndex]].count === 1) {
+			if (objFirstChar.charVal === null || objCharCounter[charArray[searchIndex]].index < objFirstChar.charVal){
```
Comment (myok12) [inline ex5B/ex5.js]: Isn't `objCharCounter[charArray[searchIndex]].index < objFirstChar.charVal` compares an index with a character?

Comment (myok12) [inline ex5B/ex5.js]: `objFirstChar.charVal === null` is only used for initialization. In this case, why did you have it initialized as `objFirstChar = { charVal: null }`? Why not just `objFirstChar = null` and changing the condition here to `if (objFirstChar === null || ...`

File: ex5B/ex5.js [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+
+   }
+
+  }
+
+  charArray = Object.keys(objCharCounter);
+
+  for (searchIndex = 0; searchIndex < Object.keys(objCharCounter).length; searchIndex++) {
+
+		if (objCharCounter[charArray[searchIndex]].count === 1) {
```
Comment (myok12) [inline ex5B/ex5.js]: Even though `objCharCounter[charArray[searchIndex]]` shouldn't take much computation, you should cache it in a variable (say, `currentCounter`) to avoid repeating this not-so-simple term many times below.

File: ex5B/ex5Spec.js [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+  });
+
+  it('finds a unique character at middle, 3 unique char', function () {
+    expect(firstUniqueChar('abcd123abcd')).toBe('1');
+  });
+
+  it('finds a unique character at the end, 3 unique char', function () {
+    expect(firstUniqueChar('abcdabcd123')).toBe('1');
+  });
+
```
Comment (myok12) [inline ex5B/ex5Spec.js]: Since we saw that the order of unique chars might be incorrectly calculated, add more tests:
`zabcd`, `abcdz`, `2b3c1a`, for example

File: ex5B/ex5.js (hunk 14 above)
Comment (zyahav) [inline ex5B/ex5.js]: Done

File: ex5B/ex5.js (hunk 13 above)
Comment (zyahav) [inline ex5B/ex5.js]: Wiil talk on the phone

File: ex5B/ex5.js (hunk 16 above)
Comment (zyahav) [inline ex5B/ex5.js]: Done

File: ex5B/ex5.js (hunk 17 above)
Comment (zyahav) [inline ex5B/ex5.js]: Done

File: ex5B/ex5.js (hunk 19 above)
Comment (zyahav) [inline ex5B/ex5.js]: Done

File: ex5B/ex5.js (hunk 18 above)
Comment (zyahav) [inline ex5B/ex5.js]: yes

File: ex5B/ex5Spec.js (hunk 20 above)
Comment (zyahav) [inline ex5B/ex5Spec.js]: Done

File: ex5B/ex5.js (hunk 18 above)
Comment (zyahav) [inline ex5B/ex5.js]: Done

Comment (zyahav) [inline ex5B/ex5.js]: Will talk about it

File: ex5B/ex5.js [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+ * For example:
+ * firstUniqueChar("abcdedcba") === "e"
+ * firstUniqueChar("abcdeedcba") === null
+ * firstUniqueChar("abababcab") === "c"
+ * firstUniqueChar("aaaaaaaaa") === null
+ * firstUniqueChar("a") === "a"
+ */
+function firstUniqueChar(str) {
+
+	var objCharCounte = buildCharIndicesMap(str);
```
Comment (myok12) [inline ex5B/ex5.js]: `objCharCounte` --> `objCharCounter`

File: ex5B/ex5.js [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+
+		}
+
+	}
+	return objCharCounter;
+}
+
+function findFirstUniqueCharInMap(objCharCounter) {
+	
+	var charArray = [], objFirstChar = { charVal: null }, currentChar, currentCounter = 0, currentFirstIndex = 0;
```
Comment (myok12) [inline ex5B/ex5.js]: As I said before, you can probably simplify the initial value you have for `objFirstChar` and have it be `var objFirstChar = null`. You'll have to change the if condition below too, obviously.
